### PR TITLE
Removes OO form matchrules, updates AS PATH

### LINKED ISSUE #14: Store matchrules as Matchrule objects and not as lists
Currently we store the list representation of `Matchrule` objects in the according `Step` and `Inspection` fields. This is a workaround that allows us to use  `attr`'s convenient `asdict` function when dumping the layout to a file. 

I suggest to [implement our own `asdict` function](https://github.com/hynek/attrs/issues/103#event-825440411) allowing us to store `Matchrules` as objects but returning their list representation instead of  their dict representation when dumping the container layout.

### PR DESCRIPTION
**Pull request by lukpueh**
- Matchrules are no longer OO
  they are simply stored as list and all the verification happens
  in verifylib
- This also fixes #14, an ugly workaround to use attr's asdict
- The commit also updates the MATCH matchrules to take an
  optional target path which allows for matching artifacts that
  have different paths

### CODE REVIEW

File: toto/verifylib.py [hunk 1]
```diff
@@ -251,47 +253,321 @@ def verify_all_steps_command_alignment(layout, links_dict):
...
-  verify them."""
+def verify_match_rule(rule, source_type, source_link, target_links):
+  """
+  <Purpose>
+    Verifies that the source artifact (depending on the list the rule was
+    extracted from a material or product) hash matches the target artifact
+    (depending on the 2nd element of the rule a material or product).
+
+    Also verifies:
+    That the target_link as idetfied by ("FROM" <step>) exists in
```
Comment (SantiagoTorres) [inline toto/verifylib.py]: identified*

File: toto/verifylib.py [hunk 2]
```diff
@@ -251,47 +253,321 @@ def verify_all_steps_command_alignment(layout, links_dict):
...
+            A dictionary of Link objects with Link names as keys.
+            The Link objects relate to Steps.
+            The contained materials and products are used as verification target.
+
+  <Exceptions>
+    raises an Exception if the rule is not conformant with the rule format.
+    raises an if a matchrule does not verify.
+    TBA (see https://github.com/in-toto/in-toto/issues/6)
+
+    RuleVerficationFailed
```
Comment (SantiagoTorres) [inline toto/verifylib.py]: When this is raised would be good saying here.

File: toto/verifylib.py [hunk 3]
```diff
@@ -251,47 +253,321 @@ def verify_all_steps_command_alignment(layout, links_dict):
...
+  # Verify that the target Link metadata object contains the material or product
+  # to match with.
+  if (target_path not in target_artifacts.keys()):
+    raise RuleVerficationFailed("'%s' not in target %ss" \
+        % (target_path, target_type))
+
+  # Verify that the recorded source artifact hash and the recorded target
+  # artifact hash are equal.
+  if (ComparableHashDict(source_artifacts[source_path]) != \
+      ComparableHashDict(target_artifacts[target_path])):
```
Comment (SantiagoTorres) [inline toto/verifylib.py:357]: Using these `ComparableHashDicts` is the way to go? couldn't we just compare fields within the source/target paths?

I wonder if this is an issue if, for example, one link provides two hashes and the other one provides only one.

This is a minor nit though.

File: toto/verifylib.py [hunk 4]
```diff
@@ -251,47 +253,321 @@ def verify_all_steps_command_alignment(layout, links_dict):
...
+    link:
+            The Link object for the Item (Step or Inspection) that contains
+            the rule.
+
+  <Exceptions>
+    raises an Exception if the rule is not conformant with the rule format.
+    raises an if a matchrule does not verify.
+    TBA (see https://github.com/in-toto/in-toto/issues/6)
+
+    RuleVerficationFailed
```
Comment (SantiagoTorres) [inline toto/verifylib.py]: same here

File: toto/verifylib.py [hunk 5]
```diff
@@ -251,47 +253,321 @@ def verify_all_steps_command_alignment(layout, links_dict):
...
+
+  <Side Effects>
+    None.
+
+  """
+  # FIXME: Validate rule format
+  path = rule[1]
+  if (path in link.materials.keys()):
+    raise RuleVerficationFailed("'%s' " \
```
Comment (SantiagoTorres) [inline toto/verifylib.py]: I don't think you need the `\` here

File: toto/verifylib.py [hunk 6]
```diff
@@ -251,47 +253,321 @@ def verify_all_steps_command_alignment(layout, links_dict):
...
+            The Link objects relate to Steps.
+            The contained materials and products are used as verification target.
+
+  <Exceptions>
+    raises an Exception if a rule is not conformant with the rule format.
+    raises an Exception if a matchrule does not verify.
+    TBA (see https://github.com/in-toto/in-toto/issues/6)
+
+  <Side Effects>
+    Calls the respective verify_<rule-type>_rule function.
```
Comment (SantiagoTorres) [inline toto/verifylib.py]: Is this really a side-effect?

File: toto/verifylib.py [hunk 7]
```diff
@@ -251,47 +253,321 @@ def verify_all_steps_command_alignment(layout, links_dict):
...
+            The Link objects relate to Steps.
+            The contained materials and products are used as verification target.
 
   <Exceptions>
     raises an Exception if a matchrule does not verify.
     TBA (see https://github.com/in-toto/in-toto/issues/6)
 
   <Side Effects>
-    Calls matchrule verification methods.
+    Calls helper functions to verify all matchrules of an item.
```
Comment (SantiagoTorres) [inline toto/verifylib.py]: Is this a side-effect?

**Review by SantiagoTorres — CHANGES_REQUESTED** [review]
Some minor nits only. Thanks!

File: toto/verifylib.py (hunk 2 above)
Comment (lukpueh) [reply toto/verifylib.py]: Okay, will give more info.

File: toto/verifylib.py (hunk 3 above)
Comment (lukpueh) [reply toto/verifylib.py:357]: `ComparableHashDicts` is just a helper I wrote to compare two hash dicts of format `ssl_crypto.formats.HASHDICT_SCHEMA`, e.g.

``` python
{'sha256': '23432df87ab..',
'sha512': '34324abc34df..', ...}
```

The behavior is documented in its docstring:

``` python
    """Equal if the dicts have the same keys and the according values
    (strings) are equal"""
```

But we can talk about changing the behavior. Using it seems handy though.

File: toto/verifylib.py (hunk 4 above)
Comment (lukpueh) [reply toto/verifylib.py]: Will change.

File: toto/verifylib.py (hunk 5 above)
Comment (lukpueh) [reply toto/verifylib.py]: Will change.

File: toto/verifylib.py (hunk 6 above)
Comment (lukpueh) [reply toto/verifylib.py]: You are right. I don't think it is a side-effect either. I guess I saw some similar "side effects" in the tuf projects and adapted the broad understanding. I'll take this out here.

File: toto/verifylib.py (hunk 7 above)
Comment (lukpueh) [reply toto/verifylib.py]: See https://github.com/in-toto/in-toto/pull/31#discussion_r86243521

File: toto/verifylib.py (hunk 3 above)
Comment (SantiagoTorres) [reply toto/verifylib.py:357]: Yeah, this was my bad. I think the current behavior is ok for the time being.

**Review by SantiagoTorres — APPROVED** [review]

**Comment by SantiagoTorres** [comment]
👍 Thanks!

**Pull request merged**